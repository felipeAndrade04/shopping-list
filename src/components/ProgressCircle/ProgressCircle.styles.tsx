import { Animated } from 'react-native';
import styled from 'styled-components/native';
import { ProgressCircleProps } from './ProgressCircle.types';

export const CircleBase = styled(Animated.View)<Partial<ProgressCircleProps>>`
  width: ${(props) => props.size}px;
  height: ${(props) => props.size}px;
  border-radius: ${(props) => props.size / 2}px;
  border-width: ${(props) => props.size / 6}px;
`;

export const EmptyCircle = styled(CircleBase)`
  border-color: ${({ theme }) => theme.colors.gray[300]};
  justify-content: center;
  align-items: center;
  transform: rotate(-45deg);
`;

export const Indicator = styled(CircleBase)`
  position: absolute;
  border-left-color: ${({ theme }) => theme.colors.dark};
  border-top-color: ${({ theme }) => theme.colors.dark};
  border-right-color: transparent;
  border-bottom-color: transparent;
`;

export const CoverIndicator = styled(CircleBase)`
  position: absolute;
  border-left-color: ${({ theme }) => theme.colors.gray[300]};
  border-top-color: ${({ theme }) => theme.colors.gray[300]};
  border-right-color: transparent;
  border-bottom-color: transparent;
`;
