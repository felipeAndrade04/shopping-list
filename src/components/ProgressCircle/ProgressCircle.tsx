import React, { useRef, useEffect } from 'react';
import { Animated } from 'react-native';
import * as S from './ProgressCircle.styles';
import { ProgressCircleProps } from './ProgressCircle.types';

export function ProgressCircle({ progress = 0, size = 50 }: ProgressCircleProps) {
  const animatedProgress = useRef(new Animated.Value(0)).current;

  const animateProgress = useRef((toValue: number) => {
    Animated.spring(animatedProgress, {
      toValue,
      useNativeDriver: false,
    }).start();
  }).current;

  useEffect(() => {
    animateProgress(progress);
  }, [animateProgress, progress]);

  const firstIndicatorRotate = animatedProgress.interpolate({
    inputRange: [0, 50],
    outputRange: ['0deg', '180deg'],
    extrapolate: 'clamp',
  });

  const secondIndicatorRotate = animatedProgress.interpolate({
    inputRange: [0, 100],
    outputRange: ['0deg', '360deg'],
    extrapolate: 'clamp',
  });

  const secondIndictorVisibility = animatedProgress.interpolate({
    inputRange: [0, 49, 50, 100],
    outputRange: [0, 0, 1, 1],
    extrapolate: 'clamp',
  });

  return (
    <S.EmptyCircle size={size}>
      <S.Indicator style={{ transform: [{ rotate: firstIndicatorRotate }] }} size={size} />
      <S.CoverIndicator size={size} />
      <S.Indicator
        size={size}
        style={{
          transform: [{ rotate: secondIndicatorRotate }],
          opacity: secondIndictorVisibility,
        }}
      />
    </S.EmptyCircle>
  );
}
