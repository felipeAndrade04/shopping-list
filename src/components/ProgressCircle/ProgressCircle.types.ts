export interface ProgressCircleProps {
  progress: number;
  size: number;
}
