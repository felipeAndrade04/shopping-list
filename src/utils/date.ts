const months = ['jan', 'fev', 'mar', 'abr', 'mai', 'jun', 'jul', 'ago', 'set', 'out', 'nov', 'dez'];

export function formatDate(date: Date) {
  const day = date.getDay();
  const month = date.getMonth();

  return `${day} de ${months[month]}`;
}
