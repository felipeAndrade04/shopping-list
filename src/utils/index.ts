export * from './delay';
export * from './date';
