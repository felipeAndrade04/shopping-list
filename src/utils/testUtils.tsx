import React from 'react';
import { authReducer } from '@app/store';
import { ReactNode } from 'react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components/native';
import * as theme from '@app/theme';
import { configureStore } from '@reduxjs/toolkit';

let store = configureStore({
  reducer: {
    auth: authReducer,
  },
});

export const wrapper = ({ children }: { children: ReactNode }) => (
  <ThemeProvider theme={theme}>
    <Provider store={store}>{children}</Provider>
  </ThemeProvider>
);

export function resetStore() {
  store = configureStore({
    reducer: {
      auth: authReducer,
    },
  });
}
