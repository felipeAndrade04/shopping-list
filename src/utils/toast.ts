import EventManager from './EvenManager';

export type ToastType = 'success' | 'info' | 'error';

export interface Toast {
  id?: string;
  type: ToastType;
  text: string;
  duration?: number;
}

export const toastEventManager = new EventManager();

export default function toast({ type, text, duration }: Toast) {
  toastEventManager.emit<Toast>('addtoast', {
    id: new Date().getTime().toString(),
    type,
    text,
    duration,
  });
}
