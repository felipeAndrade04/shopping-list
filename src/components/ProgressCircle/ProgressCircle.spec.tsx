import React from 'react';
import { render } from '@testing-library/react-native';
import { wrapper } from '@app/utils/testUtils';
import { ProgressCircle } from './ProgressCircle';

describe('<ProgressCircle />', () => {
  it('Should render correctly', () => {
    const tree = render(<ProgressCircle progress={100} size={50} />, { wrapper }).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
