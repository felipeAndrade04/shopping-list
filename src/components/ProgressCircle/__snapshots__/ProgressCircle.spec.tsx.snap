// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressCircle /> Should render correctly 1`] = `
<View
  collapsable={false}
  size={50}
  style={
    {
      "alignItems": "center",
      "borderColor": "#ACACAC",
      "borderRadius": 25,
      "borderWidth": 8.333333333333334,
      "height": 50,
      "justifyContent": "center",
      "transform": [
        {
          "rotate": "-45deg",
        },
      ],
      "width": 50,
    }
  }
>
  <View
    collapsable={false}
    size={50}
    style={
      {
        "borderBottomColor": "transparent",
        "borderLeftColor": "#0D0A0B",
        "borderRadius": 25,
        "borderRightColor": "transparent",
        "borderTopColor": "#0D0A0B",
        "borderWidth": 8.333333333333334,
        "height": 50,
        "position": "absolute",
        "transform": [
          {
            "rotate": "0deg",
          },
        ],
        "width": 50,
      }
    }
  />
  <View
    collapsable={false}
    size={50}
    style={
      {
        "borderBottomColor": "transparent",
        "borderLeftColor": "#ACACAC",
        "borderRadius": 25,
        "borderRightColor": "transparent",
        "borderTopColor": "#ACACAC",
        "borderWidth": 8.333333333333334,
        "height": 50,
        "position": "absolute",
        "width": 50,
      }
    }
  />
  <View
    collapsable={false}
    size={50}
    style={
      {
        "borderBottomColor": "transparent",
        "borderLeftColor": "#0D0A0B",
        "borderRadius": 25,
        "borderRightColor": "transparent",
        "borderTopColor": "#0D0A0B",
        "borderWidth": 8.333333333333334,
        "height": 50,
        "opacity": 0,
        "position": "absolute",
        "transform": [
          {
            "rotate": "0deg",
          },
        ],
        "width": 50,
      }
    }
  />
</View>
`;
